keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+the node has multiple network interfaces, or multiple addresses on a
+single interface). eg:
+
+    [ { "type": "ExternalIP", "address": "42.42.42.42" },
+      { "type": "InternalIP", "address": "10.1.2.3" },
+      { "type": "InternalIP", "address": "10.4.5.6" },
+      { "type": "InternalIP", "address": "10.7.8.9" } ]
+
+In this case, the primary IP would be `10.1.2.3`. If the administrator
+wants a different primary IP, they can pass `--node-ip` to choose
+another one, but that will have the side effect of removing all other
+IPs of the same type from the list. Eg, passing `--node-ip 10.4.5.6`,
+would result in `Node.Status.Addresses` having the value:
+
+    [ { "type": "ExternalIP", "address": "42.42.42.42" },
+      { "type": "InternalIP", "address": "10.4.5.6" } ]
+
+When using an external cloud-provider, the administrator has no
+control over the node IPs; in the above example, passing `--node-ip
+10.4.5.6` would have no effect on the returned list of addresses, and
+the primary IP would still be `10.1.2.3`. (Although passing `--node-ip
+10.11.12.13` would result in an error since that doesn't appear in the
+list at all.)
+
+#### API Cleanliness
+
+`Node.Status.Addresses` is annotated with an incorrect strategic-merge
+patch annotation, and so it is not possible to use strategic merge
+patch to change that field without losing information. (See discussion
+in [kubernetes #79391].) Although kubelet and cloud-controller-manager
+have been fixed to work around the problem, it is still annoying to
+have the incorrect annotation there, and it is possible that other
+components might attempt to update the field incorrectly in the
+future.
+
+Additionally, `Node.Status.Addresses` is not currently sufficiently
+checked by validation; it is possible for an external cloud provider
+to declare an `InternalIP` or `ExternalIP` address that is not a
+syntactically-valid IP address, meaning clients that read this field
+may need to sanity-check it themselves ([kubernetes #89817]). Adding
+validation now could break deployed cloud providers (which might
+otherwise appear to be fully-functional, as long as the primary IP was
+valid).
+
+[kubernetes #79391]: https://github.com/kubernetes/kubernetes/pull/79391
+[kubernetes #89817]: https://github.com/kubernetes/kubernetes/pull/89817
+
+#### `NodeAddressType` Semantics
+
+The difference between `InternalIP` and `ExternalIP` node addresses is
+not well-defined except in the case of IPv4 addresses in public
+clouds, and the two types are used inconsistently outside of that
+environment. (eg, See [kubernetes #86918] as an example of uncertainty
+about the right type to use in a particular situation.)
+
+[kubernetes #86918]: https://github.com/kubernetes/kubernetes/pull/86918#discussion_r363997750
+
+### Goals
+
+- Deprecate `Node.Status.Addresses` and replace it with new fields
+  that have clearer semantics, and do not have the strategic-merge and
+  validation problems.
```

Comment:
>     * I think we need address(es) to expose to hostNetwork=true pods via the downward API. This needs one address per family to be consistent with hostNetwork=false pods

actually, we only provide a single pod IP via the downward API right now..., but you can rewrite this point to talk about `pod.Status.PodIPs` instead

>     * I think we need address(es) that we're going to configure as targets in loadbalancers for Service type=LoadBalancer.

I'm not sure what you mean by this, particular since you say that we don't need "external IPs". LoadBalancer IPs are not, in general, expected to be associated with particular nodes (though I think in the case of MetalLB they are).

> In particular we do _not_ need a (resolvable/DNS) hostname, and we do not need "external IPs" (whatever that means).

There are a lot of references to `NodeExternalIP` in k/k. I think one case that is considered important is the case where the masters are not on the same subnet as the nodes and can't reach the internal IPs, which is sort of covered by your "kubelet endpoint address that we're going to use for execlogs calls from apiserver", except that the kubelet doesn't know whether the apiserver can reach it via InternalIP or not, so would not currently be able to set that correctly.

Reply:
> I'm not sure what you mean by this, particular since you say that we don't need "external IPs". LoadBalancer IPs are not, in general, expected to be associated with particular nodes (though I think in the case of MetalLB they are).

The underlying cloud-provider LoadBalancer needs to be configured with node addresses as 'backends' (or 'targets' or whatever the correct LB term is) - assuming the usual k8s LB->NodePort->Pod setup.  The provider code needs to know *what node address* to configure in the LB for each node (eg: in openstack provider we had to do some guesswork here because neither "internal" nor "external" clearly implied "is reachable by LB").

So not the LoadBalancer IP (VIP) itself, but the IPs that the LB load-balances _to_.

> There are a lot of references to NodeExternalIP in k/k. I think one case that is considered important is the case where the masters are not on the same subnet as the nodes and can't reach the internal IPs, which is sort of covered by your "kubelet endpoint address that we're going to use for execlogs calls from apiserver", except that the kubelet doesn't know whether the apiserver can reach it via InternalIP or not, so would not currently be able to set that correctly.

￼Oh interesting...  I see loads of references in test data, and of course in the `GetNodeHostIP()` function where they get ignored unless no internal addresses exist.

I see SSH tunnel is [hardcoded to use externalAddresses](https://github.com/kubernetes/kubernetes/blob/b93e9d9395c9de0d86fe9146f66b29329c184a25/pkg/master/master.go#L507).  Afaik we've heavily deprecated the SSH tunnel for years(?) so I don't think we care about this use case(?).

I see plenty of references in legacy-cloud-providers/ where they're generating the NodeAddresses in the first place.  My grep seems to indicate that they're all _generating_ and not _using_ the NodeExternalIP info here - but I could easily have overlooked something.

A cluster without uniform access from apiservers to kubelets would indeed be a challenge.  Where's the code you're referring to that implements this? (is this the SSH tunnel?)

I think you're also implying there are other uses of ExternalIP, beyond the SSH tunnel.  Could you point some of them out so I can learn what they're doing?  They didn't jump out in my naive NodeExternalIP grep over k/k.

At the moment I'm inclined to stand by my "external IPs are not used for anything" assuming we can ignore (or special-case) SSH tunnel.

Reply:
(So the section under discussion is gone from the document now; based on feedback, I decided to mostly give up on trying to fix Node.Status.Addresses, and I'm now only trying to fix the parts that specifically affect ipv6/dual-stack. So we're definitely not getting rid of ExternalIP in this PR now. But...)

> (eg: in openstack provider we had to do some guesswork here because neither "internal" nor "external" clearly implied "is reachable by LB")

I guess it's assumed that the LB code doesn't need to care about the _public_ node IP-related APIs, because the cloud provider is supposed to just know the cluster topology (and in theory might even route inbound LB traffic to a node IP that isn't even listed in Node.Status.Addresses).

> Afaik we've heavily deprecated the SSH tunnel for years(?) so I don't think we care about this use case(?).

Yes, it's deprecated in favor of a new thing called Konnectivity now but I didn't investigate the details.

> A cluster without uniform access from apiservers to kubelets would indeed be a challenge. Where's the code you're referring to that implements this?

You can run, eg, `kube-apiserver --kubelet-preferred-address-types ExternalDNS,ExternalIP` and then when the apiserver needs to connect to a kubelet, it will use that as the args to `nodeutil.GetPreferredNodeAddress` to decide what name/IP to connect to.